File: cpp/samples/core/codebooks.cpp
/**
 * @file codebooks.cpp
 * @brief Demonstrates usage of vista-sdk Codebooks API
 * @details This sample shows how to use VIS singleton, access codebooks,
 *          validate values, create metadata tags, and work with VIS versions
 */

#include <dnv/vista/sdk/core/VIS.h>

#include <iomanip>
#include <iostream>
#include <string>
#include <vector>

int main()
{
    using namespace dnv::vista::sdk;

    std::cout << "=== vista-sdk Codebooks Sample ===\n\n";

    {
        std::cout << "1. VIS Singleton: Version management\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& vis = VIS::instance();
        const auto& allVersions = vis.versions();

        std::cout << "Latest VIS version: " << VisVersions::toString(vis.latest()) << "\n";
        std::cout << "Available versions (" << allVersions.size() << "):\n";
        for (const auto& version : allVersions)
        {
            std::cout << "  - " << VisVersions::toString(version) << "\n";
        }
        std::cout << "\n";
    }

    {
        std::cout << "2. Codebooks: Accessing codebooks for a version\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& codebooks = VIS::instance().codebooks(VIS::instance().latest());

        std::cout << "Codebooks for version: " << VisVersions::toString(codebooks.version()) << "\n";

        const auto& quantity = codebooks[CodebookName::Quantity];
        const auto& position = codebooks[CodebookName::Position];
        const auto& state = codebooks[CodebookName::State];

        std::cout << "Quantity codebook has " << std::setw(3) << quantity.standardValues().size()
                  << " standard values\n";
        std::cout << "Position codebook has " << std::setw(3) << position.standardValues().size()
                  << " standard values\n";
        std::cout << "State    codebook has " << std::setw(3) << state.standardValues().size() << " standard values\n";
        std::cout << "\n";
    }

    {
        std::cout << "3. Codebook: Checking standard values and groups\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& quantity = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Quantity];

        std::cout << std::boolalpha;
        std::cout << "Has 'temperature'? " << quantity.hasStandardValue("temperature") << "\n";
        std::cout << "Has 'pressure'   ? " << quantity.hasStandardValue("pressure") << "\n";
        std::cout << "Has 'invalid_qty'? " << quantity.hasStandardValue("invalid_qty") << "\n";

        std::cout << "\nGroups in Quantity codebook: " << quantity.groups().size() << "\n";
        for (const auto& group : quantity.groups())
        {
            std::cout << "  - " << group << "\n";
        }
        std::cout << "\n";
    }

    {
        std::cout << "4. Position Codebook: Special numeric value handling\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& position = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Position];

        std::cout << std::boolalpha;
        std::cout << "Has '1'        ? " << position.hasStandardValue("1") << "\n";
        std::cout << "Has '03'       ? " << position.hasStandardValue("03") << "\n";
        std::cout << "Has '42'       ? " << position.hasStandardValue("42") << "\n";
        std::cout << "Has '1234'     ? " << position.hasStandardValue("1234") << "\n";
        std::cout << "Has 'centre'   ? " << position.hasStandardValue("centre") << "\n";
        std::cout << "Has 'port'     ? " << position.hasStandardValue("port") << "\n";
        std::cout << "Has 'starboard'? " << position.hasStandardValue("starboard") << "\n";
        std::cout << "Has 'invalid'  ? " << position.hasStandardValue("invalid") << "\n";
        std::cout << "\n";
    }

    {
        std::cout << "5. MetadataTag: Creating tags from standard values\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& quantity = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Quantity];

        auto tempTag = quantity.createTag("temperature");

        if (tempTag.has_value())
        {
            std::cout << "Temperature tag created:\n";
            std::cout << "  Name:        " << CodebookNames::toPrefix(tempTag->name()) << "\n";
            std::cout << "  Value:       " << tempTag->value() << "\n";
            std::cout << "  Prefix:      " << tempTag->prefix() << "\n";
            std::cout << "  Is Custom:   " << std::boolalpha << tempTag->isCustom() << "\n";

            std::string tag;
            tempTag->toString(tag);
            std::cout << "  String repr: " << tag << "\n";
        }
        std::cout << "\n";
    }

    {
        std::cout << "6. MetadataTag: Creating custom tags\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& quantity = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Quantity];

        auto customTag = quantity.createTag("custom_measurement");

        if (customTag.has_value())
        {
            std::cout << "Custom tag created:\n";
            std::cout << "  Name:        " << CodebookNames::toPrefix(customTag->name()) << "\n";
            std::cout << "  Value:       " << customTag->value() << "\n";
            std::cout << "  Prefix:      " << customTag->prefix() << "\n";
            std::cout << "  Is Custom:   " << std::boolalpha << customTag->isCustom() << "\n";

            std::string tag;
            customTag->toString(tag);
            std::cout << "  String repr: " << tag << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "7. Position Validation\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& position = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Position];

        const std::vector<std::string> testPositions = {
            "centre",
            "port",
            "1",
            "42",
            "centre-starboard",
            "port-1",
            "centre-starboard-2",
            "invalid position!",
            "starboard-centre",
        };

        std::cout << "Position validation results:\n";
        for (const auto& pos : testPositions)
        {
            auto result = position.validatePosition(pos);
            constexpr size_t columnWidth = 18;
            size_t padding = pos.length() < columnWidth ? columnWidth - pos.length() : 0;
            std::cout << "  '" << pos << "'" << std::string(padding, ' ') << ": ";

            switch (result)
            {
                case PositionValidationResult::Valid:
                    std::cout << "Valid";
                    break;
                case PositionValidationResult::Custom:
                    std::cout << "Custom";
                    break;
                case PositionValidationResult::Invalid:
                    std::cout << "Invalid";
                    break;
                case PositionValidationResult::InvalidOrder:
                    std::cout << "Invalid (wrong order)";
                    break;
                case PositionValidationResult::InvalidGrouping:
                    std::cout << "Invalid (duplicate groups)";
                    break;
            }
            std::cout << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "8. CodebookName: String conversions\n";
        std::cout << "-------------------------------------------------------\n";

        std::cout << "Enum to prefix conversions:\n";
        std::cout << "  Quantity -> '" << CodebookNames::toPrefix(CodebookName::Quantity) << "'\n";
        std::cout << "  Position -> '" << CodebookNames::toPrefix(CodebookName::Position) << "'\n";
        std::cout << "  State    -> '" << CodebookNames::toPrefix(CodebookName::State) << "'\n";
        std::cout << "  Command  -> '" << CodebookNames::toPrefix(CodebookName::Command) << "'\n";
        std::cout << "  Detail   -> '" << CodebookNames::toPrefix(CodebookName::Detail) << "'\n";

        std::cout << "Prefix to enum conversions:\n";
        std::cout << "  'qty' -> CodebookName::"
                  << (CodebookNames::fromPrefix("qty") == CodebookName::Quantity ? "Quantity" : "?") << "\n";
        std::cout << "  'pos' -> CodebookName::"
                  << (CodebookNames::fromPrefix("pos") == CodebookName::Position ? "Position" : "?") << "\n";
        std::cout << "\n";
    }

    {
        std::cout << "9. Invalid operations\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& quantity = VIS::instance().codebooks(VIS::instance().latest())[CodebookName::Quantity];

        std::cout << std::boolalpha;
        std::cout << "Empty string tag created ? " << quantity.createTag("").has_value() << "\n";
        std::cout << "Whitespace tag created   ? " << quantity.createTag("   ").has_value() << "\n";

        try
        {
            [[maybe_unused]] auto invalid = CodebookNames::fromPrefix("invalid_prefix");
            std::cout << "Invalid prefix accepted  : (unexpected)\n";
        }
        catch (const std::invalid_argument& e)
        {
            std::cout << "Invalid prefix rejected  : " << e.what() << "\n";
        }

        std::cout << "\n";
    }

    {
        std::cout << "10. Building a complete metadata path\n";
        std::cout << "-------------------------------------------------------\n";

        const auto& codebooks = VIS::instance().codebooks(VIS::instance().latest());

        auto qtyTag = codebooks[CodebookName::Quantity].createTag("temperature");
        auto cntTag = codebooks[CodebookName::Content].createTag("oil");
        auto posTag = codebooks[CodebookName::Position].createTag("1");
        auto stateTag = codebooks[CodebookName::State].createTag("running");

        if (qtyTag && cntTag && posTag && stateTag)
        {
            std::string path;

            std::cout << "Building metadata path:\n";

            qtyTag->toString(path);
            std::cout << "  After Quantity: " << path << "\n";

            cntTag->toString(path);
            std::cout << "  After Content : " << path << "\n";

            posTag->toString(path);
            std::cout << "  After Position: " << path << "\n";

            stateTag->toString(path);
            std::cout << "  After State   : " << path << "\n";
        }

        std::cout << "\n";
    }

    return 0;
}

File: bindings/rust/vista-sdk/examples/core/codebooks.rs
use vista_sdk::core::codebook::PositionValidationResult;
use vista_sdk::core::codebook_name::{codebook_names, CodebookName};
use vista_sdk::core::vis::Vis;

fn main() {
    println!("=== vista-sdk Codebooks Sample ===\n");

    {
        println!("1. VIS Singleton: Version management");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let all_versions = vis.versions();

        println!("Latest VIS version: {}", vis.latest().as_str());
        println!("Available versions ({}):", all_versions.len());
        for version in &all_versions {
            println!("  - {}", version.as_str());
        }
        println!();
    }

    {
        println!("2. Codebooks: Accessing codebooks for a version");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        println!("Codebooks for version: {}", codebooks.version().as_str());

        let quantity = &codebooks[CodebookName::Quantity];
        let position = &codebooks[CodebookName::Position];
        let state = &codebooks[CodebookName::State];

        println!(
            "Quantity codebook has {:3} standard values",
            quantity.standard_values_count()
        );
        println!(
            "Position codebook has {:3} standard values",
            position.standard_values_count()
        );
        println!(
            "State    codebook has {:3} standard values",
            state.standard_values_count()
        );
        println!();
    }

    {
        println!("3. Codebook: Checking standard values and groups");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let quantity = &codebooks[CodebookName::Quantity];

        println!(
            "Has 'temperature'? {}",
            quantity.has_standard_value("temperature")
        );
        println!(
            "Has 'pressure'   ? {}",
            quantity.has_standard_value("pressure")
        );
        println!(
            "Has 'invalid_qty'? {}",
            quantity.has_standard_value("invalid_qty")
        );

        println!("\nGroups in Quantity codebook: {}", quantity.groups_count());
        for i in 0..quantity.groups_count() {
            if let Some(group) = quantity.group_at(i) {
                println!("  - {}", group);
            }
        }
        println!();
    }

    {
        println!("4. Position Codebook: Special numeric value handling");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let position = &codebooks[CodebookName::Position];

        println!("Has '1'        ? {}", position.has_standard_value("1"));
        println!("Has '03'       ? {}", position.has_standard_value("03"));
        println!("Has '42'       ? {}", position.has_standard_value("42"));
        println!("Has '1234'     ? {}", position.has_standard_value("1234"));
        println!("Has 'centre'   ? {}", position.has_standard_value("centre"));
        println!("Has 'port'     ? {}", position.has_standard_value("port"));
        println!(
            "Has 'starboard'? {}",
            position.has_standard_value("starboard")
        );
        println!(
            "Has 'invalid'  ? {}",
            position.has_standard_value("invalid")
        );
        println!();
    }

    {
        println!("5. MetadataTag: Creating tags from standard values");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let quantity = &codebooks[CodebookName::Quantity];

        if let Some(temp_tag) = quantity.create_tag("temperature") {
            println!("Temperature tag created:");
            println!(
                "  Name:        {}",
                codebook_names::to_prefix(temp_tag.name())
            );
            println!("  Value:       {}", temp_tag.value());
            println!("  Prefix:      {}", temp_tag.prefix());
            println!("  Is Custom:   {}", temp_tag.is_custom());
            println!("  String repr: {}", temp_tag);
        }
        println!();
    }

    {
        println!("6. MetadataTag: Creating custom tags");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let quantity = &codebooks[CodebookName::Quantity];

        if let Some(custom_tag) = quantity.create_tag("custom_measurement") {
            println!("Custom tag created:");
            println!(
                "  Name:        {}",
                codebook_names::to_prefix(custom_tag.name())
            );
            println!("  Value:       {}", custom_tag.value());
            println!("  Prefix:      {}", custom_tag.prefix());
            println!("  Is Custom:   {}", custom_tag.is_custom());
            println!("  String repr: {}", custom_tag);
        }
        println!();
    }

    {
        println!("7. Position Validation");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let position = &codebooks[CodebookName::Position];

        let test_positions = [
            "centre",
            "port",
            "1",
            "42",
            "centre-starboard",
            "port-1",
            "centre-starboard-2",
            "invalid position!",
            "starboard-centre",
        ];

        println!("Position validation results:");
        for pos in &test_positions {
            let result = position.validate_position(pos);
            let label = match result {
                PositionValidationResult::Valid => "Valid",
                PositionValidationResult::Custom => "Custom",
                PositionValidationResult::Invalid => "Invalid",
                PositionValidationResult::InvalidOrder => "Invalid (wrong order)",
                PositionValidationResult::InvalidGrouping => "Invalid (duplicate groups)",
            };
            println!("  '{:<18}': {}", pos, label);
        }
        println!();
    }

    {
        println!("8. CodebookName: String conversions");
        println!("-------------------------------------------------------");

        println!("Enum to prefix conversions:");
        println!(
            "  Quantity -> '{}'",
            codebook_names::to_prefix(CodebookName::Quantity)
        );
        println!(
            "  Position -> '{}'",
            codebook_names::to_prefix(CodebookName::Position)
        );
        println!(
            "  State    -> '{}'",
            codebook_names::to_prefix(CodebookName::State)
        );
        println!(
            "  Command  -> '{}'",
            codebook_names::to_prefix(CodebookName::Command)
        );
        println!(
            "  Detail   -> '{}'",
            codebook_names::to_prefix(CodebookName::Detail)
        );

        println!("Prefix to enum conversions:");
        if let Some(name) = codebook_names::from_prefix("qty") {
            println!("  'qty' -> {}", name);
        }
        if let Some(name) = codebook_names::from_prefix("pos") {
            println!("  'pos' -> {}", name);
        }
        println!();
    }

    {
        println!("9. Invalid operations");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");
        let quantity = &codebooks[CodebookName::Quantity];

        println!(
            "Empty string tag created ? {}",
            quantity.create_tag("").is_some()
        );
        println!(
            "Whitespace tag created   ? {}",
            quantity.create_tag("   ").is_some()
        );

        match codebook_names::from_prefix("invalid_prefix") {
            None => println!("Invalid prefix rejected  : not a valid prefix"),
            Some(_) => println!("Invalid prefix accepted  : (unexpected)"),
        }
        println!();
    }

    {
        println!("10. Building a complete metadata path");
        println!("-------------------------------------------------------");

        let vis = Vis::instance();
        let codebooks = vis
            .codebooks(vis.latest())
            .expect("latest() should always be a valid VIS version");

        let qty_tag = codebooks[CodebookName::Quantity].create_tag("temperature");
        let cnt_tag = codebooks[CodebookName::Content].create_tag("oil");
        let pos_tag = codebooks[CodebookName::Position].create_tag("1");
        let state_tag = codebooks[CodebookName::State].create_tag("running");

        if let (Some(qty), Some(cnt), Some(pos), Some(state)) =
            (qty_tag, cnt_tag, pos_tag, state_tag)
        {
            println!("Building metadata path:");
            println!("  After Quantity: {}", qty);
            println!("  After Content : {}", cnt);
            println!("  After Position: {}", pos);
            println!("  After State   : {}", state);
        }
        println!();
    }
}
